src/Build/FileSystem/MSBuildFileSystemBase.cs

```diff
@@ -14,66 +15,95 @@ namespace Microsoft.Build.FileSystem
     /// - must be thread safe
     /// - may cache some or all the calls.
     /// </summary>
-    public abstract class MSBuildFileSystemBase
+    public class MSBuildFileSystemBase : IFileSystem
     {
+        private IFileSystem _defaultFileSystem;
+        private IFileSystem DefaultFileSystem
+        {
+            get
+            {
+                if (_defaultFileSystem == null)
+                {
+                    var newDefaultFileSystem = new CachingFileSystemWrapper(FileSystems.Default);
```

Comment:
Several issues with this pattern (setting a file system via the constructor and using a caching wrapper over the default if null)

A user might extend this class but not override all the methods. The methods that are not overriden will use the code in this base class which calls this DefaultFileSystem getter and will thus create a caching wrapper over msbuild's default file system implementation. This creates a partitioned logic: the user's file system which gets called part of the time, and MSBuilds cached default implementation.

Another issue is automatically choosing to cache. The user might not know this happens and they'd get a surprise memory bloating issue.

I'd keep this method non instantiable to make it clear it's a contract, not implementation.

Reply:
I like not having to implement everything, but you're right about some coherency issues if a user overrides some but not all methods. Maybe implement methods with "throw new MethodNotImplementedException()"?

Reply:
The new requirement is that users want to implement only a subset, so if MSBuild ends up calling a non implemented method, it will throw.

So one approach is the one Ladi did here, where the interface chooses some implementation of itself and uses that for non overriden methods. I'm starting to think it might be the simplest, non-bloated solution. The caveat is the forced caching which can be fixed by adding a protected switch turn caching of the default implementation on or off.

Another approach would be to split the big interface into smaller, more coherent interfaces: IExistenceChecks, IFileMetadata, IFileContents. The user can choose which one to implement and msbuild would provide default implementations for the others. But this adds a lot of boilerplate for minimal gains.

Reply:
This is a great point, thank you. How about falling back only to the non-caching default implementation, and not exposing any switches for now?

We can always add the switch later if such a need arises. Better than regretting introducing something that we later find is not needed or causing troubles. Basically we're saying that caching is internal to MSBuild and activated only if the caller didn't pass their `MSBuildFileSystemBase`-derived implementation. This way none of the caching semantics is publicly exposed.

Reply:
> keep it an abstract class to make it clear it's a contract. But this is more of a personal preference of mine, treat it as a nit :)

I hear you. I'll keep it abstract and add back the adapter as a trivial non-abstract subclass of `MSBuildFileSystemBase`.

Reply:
> How about falling back only to the non-caching default implementation, and not exposing any switches for now?

Seems the simplest. Downside is that the maximal caching benefits won't be reached. This makes me think whether it's worth supporting partial contract implementation at all. The implementor can noop the methods they don't want to support by just copying from https://github.com/dotnet/msbuild/blob/main/src/Shared/FileSystem/ManagedFileSystem.cs. Or we could make [FileSystems](https://github.com/dotnet/msbuild/blob/main/src/Shared/FileSystem/FileSystems.cs) public and the implementor can defer to `FileSystems.Default` for the things they don't want to implement.

Reply:
Yes, I was just going to reply with the same.

Although: Let's say we add a new method in a future version of MSBuild and it defaults to regular file system I/O, like the existing ones. If the goal of the 3rd party implementation is some kind of optimization, it's probably fine. Until they implement the new method they will be getting an unoptimized behavior so perhaps worst case just losing perf. If the goal is a complete virtualization of the filesystem, though, it's going to break them.

We need a mechanism to tell which version of the contract is being implemented. MSBuild can either optionally depend on a newer contract or have a hard dependency on it. The latter case would presumably generate a nice exception saying "sorry, your FS implementation is too old, I refuse to use it". I think this can be done simply with inheritance. Need to add a new method? Have `MSBuildFileSystemBase2` derive from `MSBuildFileSystemBase` so MSBuild can do an `is MSBuildFileSystemBase2` check. It's ironic that we're just trying to get rid of a similar pattern in `IBuildEngine` but it's the exact opposite scenario (interface provided vs. interface consumed) and a good fit, as far as I can tell.

The changes in this PR are then basically just for convenience so implementors that do optimizations can override only a subset of methods. Versioning would be done by introducing new types, not by adding members to existing ones. Hope it's making sense and curious what you think.

Reply:
We could have it simple: if the 3rd party implementation wants complete virtualization, they have to implement the entire interface. If they do not want complete FS virtualization but to only speedup certain calls, they can implement a subset. Caveat is, as you mentioned, that if they are going for virtualization and we introduce a new method we'll break their virtualization expectation. We could fix that by adding a version number on the interface. The implementor can check whether the latest version matches what they coded against.

From a practical viewpoint, we don't currently guarantee virtualization because the interface is not piped to all the places that do IO during evaluation. So we can just say we don't support virtualization, and hence providing default implementations for new interface members is fine. If we ever want to support virtualization and force implementors to implement newly added members, then we can add `MSBuildFileSystemBase2`.

Reply:
> From a practical viewpoint, we don't currently guarantee virtualization because the interface is not piped to all the places that do IO during evaluation.

Ah, good to know, thank you. We should aim to make it complete at some point to get the most out of the feature.

It looks like we have a reasonable path forward. I have tweaked the changes as discussed above - made `MSBuildFileSystemBase` abstract again and made it use the default FS implementation (without caching).